meilisearch-types/src/index_uid_pattern.rs

```diff
@@ -0,0 +1,125 @@
+    pub fn new_unchecked(s: impl AsRef<str>) -> Self {
+        Self(s.as_ref().to_string())
+    }
+
+    /// Matches any index name.
+    pub fn all() -> Self {
+        IndexUidPattern::from_str("*").unwrap()
+    }
+
+    /// Returns `true` if it matches any index.
+    pub fn matches_all(&self) -> bool {
+        self.0 == "*"
+    }
+
+    /// Returns `true` if the pattern matches a specific index name.
+    pub fn is_exact(&self) -> bool {
+        !self.0.ends_with('*')
+    }
+
+    /// Returns wether this index uid matches this index uid pattern.
+    pub fn matches(&self, uid: &IndexUid) -> bool {
+        self.matches_str(uid.as_str())
+    }
+
+    /// Returns wether this string matches this index uid pattern.
+    pub fn matches_str(&self, uid: &str) -> bool {
+        match self.0.strip_suffix('*') {
+            Some(prefix) => uid.starts_with(prefix),
+            None => self.0 == uid,
+        }
+    }
+}
+
+impl Deref for IndexUidPattern {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Borrow<str> for IndexUidPattern {
+    fn borrow(&self) -> &str {
+        &self.0
+    }
+}
+
+impl TryFrom<String> for IndexUidPattern {
+    type Error = IndexUidPatternFormatError;
+
+    fn try_from(uid: String) -> Result<Self, Self::Error> {
+        let result = match uid.strip_suffix('*') {
+            Some("") => Ok(IndexUidPattern(uid)),
+            Some(prefix) => IndexUid::from_str(prefix).map(|_| IndexUidPattern(uid)),
+            None => IndexUid::try_from(uid).map(IndexUid::into_inner).map(IndexUidPattern),
+        };
+
+        match result {
+            Ok(index_uid_pattern) => Ok(index_uid_pattern),
+            Err(IndexUidFormatError { invalid_uid }) => {
+                Err(IndexUidPatternFormatError { invalid_uid })
+            }
+        }
+    }
+}
+
+impl FromStr for IndexUidPattern {
+    type Err = IndexUidPatternFormatError;
+
+    fn from_str(uid: &str) -> Result<IndexUidPattern, IndexUidPatternFormatError> {
+        uid.to_string().try_into()
+    }
+}
+
+impl From<IndexUidPattern> for String {
+    fn from(IndexUidPattern(uid): IndexUidPattern) -> Self {
+        uid
+    }
+}
+
+#[derive(Debug)]
+pub struct IndexUidPatternFormatError {
+    pub invalid_uid: String,
+}
+
+impl fmt::Display for IndexUidPatternFormatError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "`{}` is not a valid index uid pattern. Index uid patterns \
+            can be an integer or a string containing only alphanumeric \
+            characters, hyphens (-), underscores (_), and \
+            optionally end with a star (*).",
+            self.invalid_uid,
+        )
+    }
+}
+
+impl Error for IndexUidPatternFormatError {}
+
+impl ErrorCode for IndexUidPatternFormatError {
+    fn error_code(&self) -> Code {
+        // TODO should I return a new error code?
+        Code::InvalidIndexUid
```

Comment:
here it's no more an IndexUid but an IndexUidPattern, so I think it would be better to have a new error code. 🤔 (poke @gmourier)

Reply:
Personally, I don't think this error code matter that much for the end user, but I may be wrong :thinking:

Reply:
Here is what we have for the swap indexes resource when you give a wrong pattern for an index uid in the index array.

We could follow the same rule to stay consistent.

https://github.com/meilisearch/specifications/blob/main/text/0191-swap-indexes-api.md#34131-synchronous-errors

> Sending an indexes array containing an invalid index uid format returns an [invalid_swap_indexes](https://github.com/meilisearch/specifications/blob/main/text/0061-error-format-and-definitions.md#invalid_swap_indexes) error.

It means keeping `invalid_api_key_indexes` for the error code and injecting the `invalid_index_uid` variant for the `message` for the sake of consistency.

Reply:
@ManyTheFish @irevoire, the `IndexUid` struct was returning an `InvalidIndexUid`. I think I should keep what I wrote here and also return this `Code` for when an `IndexUidPattern` is wrong, right? Doing that will keep the consistency of the API and the error message is custom and really easy to understand.

Reply:
@Kerollmops, I'm not sure to understand.
Now you always take an `IndexUidPattern` in parameter. No matter if there is a `*` at the end or not, it's an `IndexUidPattern` no? 🤔

Reply:
After a quick call with @ManyTheFish we decided that we will keep the error code I currently use: `Code::InvalidIndexUid`. The error message is sent by Meilisearch and is clear enough for a user to understand. Thank you everybody 😍